src/format_rules.ml

```diff
@@ -0,0 +1,81 @@
+open Import
+
+let flag_of_kind : Ml_kind.t -> _ =
+  function
+  | Impl -> "--impl"
+  | Intf -> "--intf"
+
+let config_includes (config : Dune_file.Auto_format.t) s =
+  match config.enabled_for with
+  | Default -> true
+  | Only set -> List.mem s ~set
+
+let add_diff sctx loc ~dir input output =
+  let module SC = Super_context in
+  let open Build.O in
+  let action = Action.diff input output in
+  let alias_conf_name = "fmt" in
+  let stamp = ("format-alias", action) in
+  let alias = Build_system.Alias.make alias_conf_name ~dir in
+  SC.add_alias_action sctx alias ~loc:(Some loc) ~locks:[] ~stamp
+    (Build.paths [input; output]
+     >>>
+     Build.action
+       ~dir
+       ~targets:[]
+       action)
+
+let gen_rules sctx (config : Dune_file.Auto_format.t) ~dir =
+  let loc = config.loc in
+  let files =
+    File_tree.files_of
+      (Super_context.file_tree sctx)
+      (Path.drop_build_context_exn dir)
+  in
+  let programs = Hashtbl.create 3 in
+  let resolve_program program_name =
+    Hashtbl.find_or_add
```

Comment:
It feels like this caching should go directly in `Super_context.resolve_program`. In particular, we resolve `refmt` in `preprocessing.ml` as well.

Reply:
Done. On the test suite, the cache hit rate is 63%.